src/transform/src/dataflow.rs

```diff
@@ -167,6 +167,72 @@ fn inline_views(dataflow: &mut DataflowDesc) {
             .transform(object.relation_expr.as_mut(), &indexes)
             .unwrap();
     }
+
+    prune_unused_imported_indexes(dataflow, &indexes);
+}
+
+/// Finds the imported indexes that are actually used by any arrangement in the
+/// dataflow and remove the rest from the collection of imported ones.
+fn prune_unused_imported_indexes(
+    dataflow: &mut DataflowDesc,
+    indexes_by_collection: &HashMap<GlobalId, Vec<(GlobalId, Vec<MirScalarExpr>)>>,
+) {
+    // Traverse the dataflow to find the imported indexes that are actually used
+    let mut used_indexes = HashSet::new();
+    for object in dataflow.objects_to_build.iter_mut() {
+        object.relation_expr.as_mut().visit_mut_pre(&mut |x| {
```

Comment:
Totally random question, since you are not mutating the object as you walk it, do you need all the "mut" stuff?

Reply:
I guess I could use a different non-mutable visitor method.

One thing I'd like thought, and I haven't found, is a walker method that allows you to prune part of the subgraph. For example, once we have matched the current node against the nested pattern `ArrangeBy -> Get` we don't need to visit the nested `Get` operator.

Reply:
In `src/expr/src/scalar/mod.rs` (a.k.a. the implementation for `MirScalarExpr`, there is a method `visit_mut_pre_post` that allows you specify whether or not you want to keep recursing down. See if adding similar method to `MirRelationExpr` would work.